node/pkg/evm/connectors/poller.go

```diff
@@ -0,0 +1,218 @@
+	errFeed   ethEvent.Feed
+}
+
+func NewBlockPollConnector(ctx context.Context, baseConnector Connector, finalizer PollFinalizer, delayInMs int, isEthPoS bool) (*BlockPollConnector, error) {
+	connector := &BlockPollConnector{
+		Connector:           baseConnector,
+		DelayInMs:           delayInMs,
+		isEthPoS:            isEthPoS,
+		hasEthSwitchedToPoS: false,
+		finalizer:           finalizer,
+	}
+	err := supervisor.Run(ctx, "blockPoller", connector.run)
+	if err != nil {
+		return nil, err
+	}
+	return connector, nil
+}
+
+func (b *BlockPollConnector) run(ctx context.Context) error {
+	logger := supervisor.Logger(ctx).With(zap.String("eth_network", b.Connector.NetworkName()))
+
+	latestBlock, err := b.getBlock(ctx, logger, nil)
+	if err != nil {
+		return err
+	}
+	currentBlockNumber := new(big.Int).Set(latestBlock.Number)
+
+	timer := time.NewTimer(time.Millisecond) // Start immediately.
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	var errorCount int
+	for {
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return ctx.Err()
+		case <-timer.C:
+			var errorOccurred bool
+			for {
+				var block *NewBlock
+				var err error
+				errorOccurred = false
+
+				// See if the next block has been created yet.
+				if currentBlockNumber.Cmp(latestBlock.Number) > 0 {
+					tmpLatestBlock, latestBlockErr := b.getBlock(ctx, logger, nil)
+					if latestBlockErr != nil {
+						errorOccurred = true
+						logger.Error("failed to look up latest block",
+							zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(latestBlockErr))
+						break
+					}
+					latestBlock = tmpLatestBlock
+
+					if currentBlockNumber.Cmp(latestBlock.Number) > 0 {
+						// We have to wait for this block to become available.
+						break
+					}
+
+					if currentBlockNumber.Cmp(latestBlock.Number) == 0 {
+						block = latestBlock
+					}
+				}
+
+				// Fetch the hash every time, in case it changes due to a rollback. The only exception is if we just got it above.
+				if block == nil {
+					block, err = b.getBlock(ctx, logger, currentBlockNumber)
+					if err != nil {
+						errorOccurred = true
+						logger.Error("failed to get current block",
+							zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(err))
+						break
+					}
+				}
+
+				if b.finalizer != nil {
+					finalized, err := b.finalizer.IsBlockFinalized(ctx, block)
+					if err != nil {
+						errorOccurred = true
+						logger.Error("failed to see if block is finalized",
+							zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(err))
+						break
+					}
+
+					if !finalized {
+						break
+					}
+				}
+
+				b.blockFeed.Send(block)
+				currentBlockNumber.Add(currentBlockNumber, big.NewInt(1))
+			}
+
+			if errorOccurred {
+				errorCount++
+				if errorCount > 1 {
+					b.errFeed.Send(fmt.Errorf("polling encountered multiple errors"))
+				}
+			} else {
+				errorCount = 0
+			}
+
+			timer = time.NewTimer(time.Duration(b.DelayInMs) * time.Millisecond)
```

Comment:
I think this is fine the way it is but we know the timer has expired and we've drained the channel (because we're in the `case <-timer.C` block) so we could also do `timer.Reset(time.Duration(b.DelayInMs) * time.Millisecond)`